sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1229,3 +1229,140 @@ case class Flatten(child: Expression) extends UnaryExpression {
+        leftGen.code +
+          rightGen.code + ctx.nullSafeExec(right.nullable, rightGen.isNull) {
+            s"""
+              ${ev.isNull} = false;
+              $resultCode
+            """
+          }
+
+      ev.copy(code =
+        s"""
+           | boolean ${ev.isNull} = true;
+           | ${CodeGenerator.javaType(dataType)} ${ev.value} =
+           |   ${CodeGenerator.defaultValue(dataType)};
+           | $nullSafeEval
+         """.stripMargin
+      )
+    } else {
+      ev.copy(code =
+        s"""
+           | boolean ${ev.isNull} = false;
+           | ${leftGen.code}
+           | ${rightGen.code}
+           | ${CodeGenerator.javaType(dataType)} ${ev.value} =
+           |   ${CodeGenerator.defaultValue(dataType)};
+           | $resultCode
+         """.stripMargin
+        , isNull = FalseLiteral)
+    }
+
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+
+    nullSafeCodeGen(ctx, ev, (l, r) => {
+      val et = dataType.elementType
+      val isPrimitive = CodeGenerator.isPrimitiveType(et)
+
+      val arrayDataName = ctx.freshName("arrayData")
+      val arrayName = ctx.freshName("arrayObject")
+      val numElements = ctx.freshName("numElements")
+
+      val genNumElements =
+        s"""
+           | int $numElements = 0;
+           | if ($r > 0) {
+           |   $numElements = $r;
+           | }
+         """.stripMargin
+
+      val initialization = if (isPrimitive) {
+        val arrayName = ctx.freshName("array")
+        val baseOffset = Platform.BYTE_ARRAY_OFFSET
+        s"""
+           | int numBytes = ${et.defaultSize} * $numElements;
+           | int unsafeArraySizeInBytes =
+           |   UnsafeArrayData.calculateHeaderPortionInBytes($numElements)
+           |     + org.apache.spark.unsafe.array.ByteArrayMethods
+           |       .roundNumberOfBytesToNearestWord(numBytes);
+           | byte[] $arrayName = new byte[unsafeArraySizeInBytes];
+           | UnsafeArrayData $arrayDataName = new UnsafeArrayData();
+           | Platform.putLong($arrayName, $baseOffset, $numElements);
+           | $arrayDataName.pointTo($arrayName, $baseOffset, unsafeArraySizeInBytes);
+           | ${ev.value} = $arrayDataName;
+         """.stripMargin
+      } else {
+        s"${ev.value} = new ${classOf[GenericArrayData].getName()}(new Object[$numElements]);"
+      }
+
+      val primitiveValueTypeName = CodeGenerator.primitiveTypeName(et)
+      val assignments = {
+        val updateArray = if (isPrimitive) {
+          val isNull = left.genCode(ctx).isNull
+          s"""
+             | if ($isNull) {
+             |   ${ev.value}.setNullAt(k);
+             | } else {
+             |   ${ev.value}.set$primitiveValueTypeName(k, $l);
+             | }
+           """.stripMargin
+        } else {
+          s"${ev.value}.update(k, $l);"
+        }
+        s"""
+           | for (int k = 0; k < $numElements; k++) {
+           |   ${updateArray};
+           | }
```

Comment:
Maybe we should check if null and then loop to fill the values to avoid null-check each time.

```scala
if (left is null) {
  for (...) {
    set nulls
  }
} else {
  for (...) {
    set values
  }
}
```

if we use this, we don't need to set nulls for non-primitive types because `new Object[]` creates an array containing only `null`s.